### PR TITLE
feat: add springboot and micronaut grpc bff example

### PR DESCRIPTION
**Pull request by luismaximozup**
add bff examples using spring-boot and micronaut grpc application with beagle

### CODE REVIEW

**Comment by botbeaglezup** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: micronaut-grpc-bff/README.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+### How to execute?
+
+- Requirements:
+    - Open JDK 11
+  
+
+- Now, just open your terminal at the root of the project and run ./gradlew run
+
+Ready! Application will be standing at door 50051, enjoy it!
```
Comment (LG95) [inline micronaut-grpc-bff/README.md:13]:
```suggestion
That's it! The application will be up at localhost:50051, enjoy it!
```

File: micronaut-grpc-bff/README.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+### Services available:
+
+- ScreenService
+  - method: getScreen
+  - request: ```{
+      "name": "String",  
+      "parameters": "String"  
+    }```
+    - Description: This method searches for one or more Screen Builders registered in the BeagleScreenService class provided by lib.   
+      - **"name"** parameter is mandatory and must contain a valid string referring to the name of a registered screen.
```
Comment (LG95) [inline micronaut-grpc-bff/README.md:24]:
```suggestion
      - **"name"** is a mandatory parameter and must contain a valid string referring to the name of a registered screen. For this application, the available names are "Home" and "Text".
```

File: micronaut-grpc-bff/README.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+- ScreenService
+  - method: getScreen
+  - request: ```{
+      "name": "String",  
+      "parameters": "String"  
+    }```
+    - Description: This method searches for one or more Screen Builders registered in the BeagleScreenService class provided by lib.   
+      - **"name"** parameter is mandatory and must contain a valid string referring to the name of a registered screen.  
+      - **"Parameters"** is optional and contains a string that is used in the message of the Text component of the Home Screen
```
Comment (LG95) [inline micronaut-grpc-bff/README.md:25]:
```suggestion
      - **`parameters`** is optional and contains a string that is used in the message of the `Text` component of the home screen
```

File: micronaut-grpc-bff/gradle.properties [hunk 4]
```diff
@@ -0,0 +1,2 @@
+micronautVersion=2.4.2
+kotlinVersion=1.4.31
```
Comment (LG95) [inline micronaut-grpc-bff/gradle.properties:2]: Why not place the version of all dependencies here?

Comment (luismaximozup) [reply micronaut-grpc-bff/gradle.properties:2]: Ok

File: spring-grpc-bff/README.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+- ScreenService
+    - method: getScreen
+    - request: ```{
+      "name": "String",  
+      "parameters": "String"  
+      }```
+        - Description: This method searches for one or more Screen Builders registered in class BeagleScreenService provided by lib.
+            - **"name"** parameter is mandatory and must contain a valid string referring the name of a registered screen.
+            - **"Parameters"** is optional and contains a string that is used in Text component message on Home Screen
```
Comment (LG95) [inline spring-grpc-bff/README.md:25]:
```suggestion
That's it! The application will be up at localhost:50051, enjoy it!

### Services available:

- ScreenService
  - method: getScreen
  - request: ```{
      "name": "String",  
      "parameters": "String"  
    }```
    - Description: This method searches for one or more Screen Builders registered in the BeagleScreenService class provided by lib.   
      - **`name`** is a mandatory parameter and must contain a valid string referring to the name of a registered screen. For this application, the available names are "Home" and "Text".
      - **`parameters`** is optional and contains a string that is used in the message of the `Text` component of the home screen
```

**Review by LG95 — CHANGES_REQUESTED** [review]
Why is gradle wrapper added to the project's root and `grpc-backend-lib` in this branch?

**Comment by luismaximozup** [comment]
> Why is gradle wrapper added to the project's root and `grpc-backend-lib` in this branch?

Because another PR will be opened refactoring and joining within the backend. If everything was done together initially, it would be a big PR making the review process more difficult.

**Comment by LG95** [comment]
> > Why is gradle wrapper added to the project's root and `grpc-backend-lib` in this branch?
> 
> Because another PR will be opened refactoring and joining within the backend. If everything was done together initially, it would be a big PR making the review process more difficult.

I see. Truly larger PRs are harder to review, but when splitting the change across branches, make sure they are independent, otherwise all the changes need to be analyzed together and might as well have been a singe PR.
This explains the `grpc-backend-lib` directory, but what about the addition of gradle wrapper to the root?

File: micronaut-grpc-bff/src/main/kotlin/br/com/zup/interceptor/MyInterceptor.kt [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package br.com.zup.interceptor
+
+import br.com.zup.grpc.BeagleHeadersInterceptor
```
Comment (LG95) [inline micronaut-grpc-bff/src/main/kotlin/br/com/zup/interceptor/MyInterceptor.kt:3]: Where is this defined? I can't find it in any file in this PR nor in #65

File: spring-grpc-bff/src/main/kotlin/br/com/zup/grpcbff/service/MyBeagleScreenService.kt [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+import br.com.zup.grpcbff.interceptor.BeagleHeadersInterceptor
+import br.com.zup.grpc.BeagleScreenService
+import net.devh.boot.grpc.server.service.GrpcService
+import br.com.zup.grpcbff.screen.HomeScreen
+import br.com.zup.grpcbff.screen.SplashScreen
+
+@GrpcService(interceptors = [BeagleHeadersInterceptor::class])
+class MyBeagleAppService: BeagleScreenService() {
+    override fun screens(): Map<String, (String) -> ScreenBuilder> {
+        val requestHeaders = BeagleHeadersInterceptor.HEADERS_VALUE.get()
```
Comment (LG95) [inline spring-grpc-bff/src/main/kotlin/br/com/zup/grpcbff/service/MyBeagleScreenService.kt:13]: Why not create a screen that use these headers for a more complete example?

File: micronaut-grpc-bff/src/main/kotlin/br/com/zup/interceptor/MyInterceptor.kt (hunk 6 above)
Comment (luismaximozup) [reply micronaut-grpc-bff/src/main/kotlin/br/com/zup/interceptor/MyInterceptor.kt]: done

File: spring-grpc-bff/src/main/kotlin/br/com/zup/grpcbff/service/MyBeagleScreenService.kt (hunk 7 above)
Comment (luismaximozup) [reply spring-grpc-bff/src/main/kotlin/br/com/zup/grpcbff/service/MyBeagleScreenService.kt:13]: nice suggestion, I will think of something

File: micronaut-grpc-bff/src/main/kotlin/br/com/zup/interceptor/ServerInterceptorFactory.kt [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+import io.micronaut.context.annotation.Factory
+import io.micronaut.grpc.server.interceptor.OrderedServerInterceptor
+import javax.inject.Singleton
+
+@Factory
+class ServerInterceptorFactory {
+    @Bean
+    @Singleton
+    fun mySeverInterceptor(): ServerInterceptor {
+        return  OrderedServerInterceptor ( MyInterceptor(), 10 ); // (3)
```
Comment (LG95) [inline micronaut-grpc-bff/src/main/kotlin/br/com/zup/interceptor/ServerInterceptorFactory.kt:14]: What is this comment about?

Comment (luismaximozup) [reply micronaut-grpc-bff/src/main/kotlin/br/com/zup/interceptor/ServerInterceptorFactory.kt]: fixed

File: micronaut-grpc-bff/src/main/kotlin/br/com/zup/screen/SplashScreen.kt [hunk 9]
```diff
@@ -37,7 +37,7 @@ object SplashScreen : ScreenBuilder {
 
                 buttonWithAppearanceAndStyle(text = "Button with Appearance"),
                 buttonWithAppearanceAndStyle(
-                    text = "Button with Appearance and style",
+                    text = "Button with Appearance, style and headers: ${this.headers["content-type"]}",
```
Comment (LG95) [inline micronaut-grpc-bff/src/main/kotlin/br/com/zup/screen/SplashScreen.kt:40]:
```suggestion
                    text = "Button with Appearance, style and content type: ${this.headers["content-type"]}",
```

File: spring-grpc-bff/src/main/kotlin/br/com/zup/grpcbff/screen/SplashScreen.kt [hunk 10]
```diff
@@ -37,7 +37,7 @@ object SplashScreen : ScreenBuilder {
 
                 buttonWithAppearanceAndStyle(text = "Button with Appearance"),
                 buttonWithAppearanceAndStyle(
-                    text = "Button with Appearance and style",
+                    text = "Button with Appearance, style and headers: ${this.headers["content-type"]}",
```
Comment (LG95) [inline spring-grpc-bff/src/main/kotlin/br/com/zup/grpcbff/screen/SplashScreen.kt:40]:
```suggestion
                    text = "Button with Appearance, style and content type: ${this.headers["content-type"]}",
```

**Review by LG95 — APPROVED** [review]

**Review by LG95 — APPROVED** [review]

**Pull request merged**